ocaml/xapi/storage_access.ml

```diff
@@ -1022,22 +1007,60 @@ let on_xapi_start ~__context =
     end
     else smapiv2_drivers
   in
+  (* Add all the running SMAPIv2 drivers *)
+  let to_keep = to_keep @ running_smapiv2_drivers in
+
+  (* Delete all records which aren't configured or in-use *)
+  List.iter
+    (fun ty ->
+       info "Unregistering SM plugin %s since not in the whitelist and not in-use" ty;
+       let self, _ = List.assoc ty existing in
+       try
+         Db.SM.destroy ~__context ~self
+       with _ -> ()
+    ) (List.set_difference (List.map fst existing) to_keep);
+
+  (* Synchronize SMAPIv1 plugins *)
+
+  (* Create all missing SMAPIv1 plugins *)
+  List.iter
+    (fun ty ->
+       let query_result = Sm.info_of_driver ty |> Smint.query_result_of_sr_driver_info in
+       Xapi_sm.create_from_query_result ~__context query_result
+    ) (List.set_difference smapiv1_drivers (List.map fst existing));
+  (* Update all existing SMAPIv1 plugins *)
+  List.iter
+    (fun ty ->
+       let query_result = Sm.info_of_driver ty |> Smint.query_result_of_sr_driver_info in
+       Xapi_sm.update_from_query_result ~__context (List.assoc ty existing) query_result
+    ) (List.intersect smapiv1_drivers (List.map fst existing));
+
+  (* Synchronize SMAPIv2 plugins *)
+
   (* Create all missing SMAPIv2 plugins *)
-  let query ty =
-    let queue_name = !Storage_interface.queue_name ^ "." ^ ty in
-    let uri () = Storage_interface.uri () ^ ".d/" ^ ty in
-    let rpc = external_rpc queue_name uri in
-    let module C = Storage_interface.Client(struct let rpc = rpc end) in
-    let dbg = Context.string_of_task __context in
-    C.Query.query ~dbg in
+  let with_query_result ty f =
+    let query ty =
+      let queue_name = !Storage_interface.queue_name ^ "." ^ ty in
+      let uri () = Storage_interface.uri () ^ ".d/" ^ ty in
+      let rpc = external_rpc queue_name uri in
+      let module C = Storage_interface.Client(struct let rpc = rpc end) in
+      let dbg = Context.string_of_task __context in
+      C.Query.query ~dbg
+    in
+    try
```

Comment:
Couldn't we use log_and_ignore_exn here? We would see the stacktrace in the log in that case (and the stacktrace might be a combined ocaml+python one if everything works correctly)

Reply:
Good idea, done.